File: test/zig/walk_order_test.zig
//! Verifies the walker enters every parent's children in source order.
//!
//! Node payload structs in `ast.zig` declare their child fields in source
//! order and the walker visits struct fields in declaration order, so this
//! sweep over the full `test/parser` corpus catches any field that drifts
//! out of order.

const std = @import("std");
const parser = @import("parser");
const utils = @import("utils.zig");

const ast = parser.ast;
const traverser = parser.traverser;
const Allocator = std.mem.Allocator;

const Ctx = struct { tree: *const ast.Tree };

const NodeTag = std.meta.Tag(ast.NodeData);

const Frame = struct {
    tag: NodeTag,
    last: u32 = 0,
    /// quasis and expressions interleave in source, but walk order is by
    /// field, matching every ESTree walker
    interleaved: bool,
};

const Violation = struct {
    parent: NodeTag,
    child: NodeTag,
    start: u32,
    prev: u32,
};

const OrderVisitor = struct {
    gpa: Allocator,
    frames: std.ArrayList(Frame) = .empty,
    violation: ?Violation = null,

    pub fn enter_node(
        self: *OrderVisitor,
        data: ast.NodeData,
        index: ast.NodeIndex,
        ctx: *Ctx,
    ) Allocator.Error!traverser.Action {
        const start = ctx.tree.span(index).start;
        if (self.frames.items.len > 0) {
            const frame = &self.frames.items[self.frames.items.len - 1];
            if (!frame.interleaved) {
                if (start < frame.last) {
                    self.violation = .{
                        .parent = frame.tag,
                        .child = data,
                        .start = start,
                        .prev = frame.last,
                    };
                    return .stop;
                }
                frame.last = start;
            }
        }
        try self.frames.append(self.gpa, .{
            .tag = data,
            .interleaved = data == .template_literal or data == .ts_template_literal_type,
        });
        return .proceed;
    }

    pub fn exit_node(
        self: *OrderVisitor,
        data: ast.NodeData,
        index: ast.NodeIndex,
        ctx: *Ctx,
    ) void {
        _ = data;
        _ = index;
        _ = ctx;
        _ = self.frames.pop();
    }
};

const OrderChecker = struct {
    gpa: Allocator,

    pub fn check(self: OrderChecker, path: []const u8, tree: *const ast.Tree) !void {
        var visitor: OrderVisitor = .{ .gpa = self.gpa };
        defer visitor.frames.deinit(self.gpa);
        var ctx: Ctx = .{ .tree = tree };
        try traverser.walk(Ctx, OrderVisitor, &visitor, &ctx);

        if (visitor.violation) |v| {
            std.debug.print(
                "{s}: {t} at offset {d} entered after a sibling at {d} inside {t}\n",
                .{ path, v.child, v.start, v.prev, v.parent },
            );
            return error.ChildrenOutOfOrder;
        }
    }
};

test "walker enters children in source order across the parser corpus" {
    const gpa = std.testing.allocator;
    try utils.forEachCorpusTree(gpa, OrderChecker{ .gpa = gpa });
}

File: test/ast/walk-order.test.ts
import { describe, expect, test } from "bun:test";
import type { Node } from "yuku-parser";
import { parse } from "yuku-parser";
import { walk } from "yuku-ast";
import { corpusFiles, corpusPresent } from "../corpus";

// quasis and expressions interleave in source, but walk order is by
// key, matching every ESTree walker
const INTERLEAVED_TYPES = new Set(["TemplateLiteral", "TSTemplateLiteralType"]);

/** The first child entered out of source order, if any. */
function walkOrderViolation(program: Node): string | null {
  let violation: string | null = null;
  const lastStart = new Map<Node, number>();
  walk(program, {
    enter(node, ctx) {
      const parent = ctx.parent;
      if (parent === null || INTERLEAVED_TYPES.has(parent.type)) return;
      const prev = lastStart.get(parent);
      if (prev !== undefined && node.start < prev) {
        violation = `${node.type} at ${node.start} entered after a sibling at ${prev} inside ${parent.type}`;
        ctx.stop();
        return;
      }
      lastStart.set(parent, node.start);
    },
  });
  return violation;
}

describe("walk order", () => {
  test("siblings are entered in source order", () => {
    const { program } = parse(`const a = f(1, x + y, [b, ...c]); class C { m() {} n() {} }`, {
      sourceType: "module",
      lang: "ts",
    });
    expect(walkOrderViolation(program)).toBeNull();
  });

  test.skipIf(!corpusPresent())("holds across the parser corpus", async () => {
    const violations: string[] = [];
    let checked = 0;
    for (const file of corpusFiles()) {
      const source = await Bun.file(file.path).text();
      const { program, diagnostics } = parse(source, {
        sourceType: file.sourceType,
        lang: file.lang,
      });
      if (diagnostics.length > 0) continue;
      checked++;
      const violation = walkOrderViolation(program);
      if (violation !== null && violations.length < 8) {
        violations.push(`${file.path}: ${violation}`);
      }
    }
    console.log(`walk order: ${checked} files in source order`);
    expect(violations).toEqual([]);
    expect(checked).toBeGreaterThan(1000);
  }, 240_000);
});
